src/services/storage-service.js

```diff
@@ -134,27 +134,65 @@ export default class StorageService {
+
+    let currentDirectory;
+    if (currentDirectorySegments.length > 0) {
+      currentDirectory = {
+        displayName: currentDirectorySegments[currentDirectorySegments.length - 1],
+        parentPath: currentDirectorySegments.slice(0, -1).join('/'),
+        path: currentDirectorySegments.join('/'),
+        url: [this.serverConfig.cdnRootUrl, ...encodedCurrentDirectorySegments].join('/'),
+        portableUrl: `cdn://${encodedCurrentDirectorySegments.join('/')}`,
+        createdOn: null,
+        type: CDN_OBJECT_TYPE.directory,
+        size: null
+      };
+    } else {
+      currentDirectory = null;
+    }
+
+    let parentDirectory;
+    if (currentDirectorySegments.length > 1) {
+      const parentDirectorySegments = currentDirectorySegments.slice(0, -1);
+      const encodedParentDirectorySegments = encodedCurrentDirectorySegments.slice(0, -1);
+
+      parentDirectory = {
+        displayName: parentDirectorySegments[parentDirectorySegments.length - 1],
+        parentPath: parentDirectorySegments.slice(0, -1).join('/'),
+        path: parentDirectorySegments.join('/'),
+        url: [this.serverConfig.cdnRootUrl, ...encodedParentDirectorySegments].join('/'),
+        portableUrl: `cdn://${encodedParentDirectorySegments.join('/')}`,
+        createdOn: null,
+        type: CDN_OBJECT_TYPE.directory,
+        size: null
+      };
+    } else {
+      parentDirectory = null;
+    }
+
+    const prefix = currentDirectorySegments.length ? `${currentDirectorySegments.join('/')}/` : '';
+    const cdnObjects = await this.cdn.listObjects({ prefix, recursive });
 
-    const mappedObjects = objects.map(obj => {
+    const objects = cdnObjects.map(obj => {
       const isDirectory = !!obj.prefix;
       const path = isDirectory ? obj.prefix : obj.name;
-      const segments = path.split('/').filter(seg => !!seg);
-      const encodedSegments = segments.map(s => encodeURIComponent(s));
+      const objectSegments = path.split('/').filter(seg => !!seg);
+      const encodedObjectSegments = objectSegments.map(s => encodeURIComponent(s));
       return {
-        displayName: segments[segments.length - 1],
-        parentPath: segments.slice(0, -1).join('/'),
-        path: segments.join('/'),
-        url: [this.serverConfig.cdnRootUrl, ...encodedSegments].join('/'),
-        portableUrl: `cdn://${encodedSegments.join('/')}`,
+        displayName: objectSegments[objectSegments.length - 1],
+        parentPath: objectSegments.slice(0, -1).join('/'),
+        path: objectSegments.join('/'),
+        url: [this.serverConfig.cdnRootUrl, ...encodedObjectSegments].join('/'),
+        portableUrl: `cdn://${encodedObjectSegments.join('/')}`,
         createdOn: isDirectory ? null : obj.lastModified,
         type: isDirectory ? CDN_OBJECT_TYPE.directory : CDN_OBJECT_TYPE.file,
         size: isDirectory ? null : obj.size
       };
     });
 
-    return mappedObjects;
+    return { parentDirectory, currentDirectory, objects };
   }
```

Comment:
This method is decomposing and recomposing paths alot, feels like a test would be helpful here 🤔

Reply:
True, it's the one function in this service that has no tests, I will add some.

Reply:
Added, thanks for the suggestion, I actually found some edge case problems.